R-package/src/cmake/modules/FindLibR.cmake

```diff
@@ -174,17 +174,19 @@ execute_process(
   OUTPUT_VARIABLE LIBR_LIB_DIR
 )
 
+set(LIBR_HOME ${LIBR_HOME} CACHE PATH "R home directory")
+set(LIBR_EXECUTABLE ${LIBR_EXECUTABLE} CACHE PATH "R executable")
+set(LIBR_INCLUDE_DIRS ${LIBR_INCLUDE_DIRS} CACHE PATH "R include directory")
+set(LIBR_LIB_DIR ${LIBR_LIB_DIR} CACHE PATH "R shared libraries directory")
+
 # look for the core R library
 find_library(
   LIBR_CORE_LIBRARY
-  NAMES R
-  HINTS "${CMAKE_CURRENT_BINARY_DIR}" "${LIBR_LIB_DIR}" "${LIBR_HOME}/bin" "${LIBR_LIBRARIES}"
+  NAMES R R.dll
+  HINTS "${CMAKE_CURRENT_BINARY_DIR}" "${LIBR_HOME}/lib" "${LIBR_HOME}/bin/${R_ARCH}" "${LIBR_HOME}/bin" "${LIBR_LIBRARIES}"
+  PATHS "${LIBR_HOME}/bin/${R_ARCH}/R.dll"
 )
```

Comment:
@StrikerRUS I'm starting a thread here for [this discussion](https://github.com/microsoft/LightGBM/pull/2963#discussion_r401164197) on how to find `R.dll`.

Reply:
to summarize for reviewers:

`find_library()` doesn't consider `.dll` by default ([link 1](http://cmake.3232098.n2.nabble.com/find-library-doesn-t-find-dll-on-windows-td7597643.html), [link 2](https://stackoverflow.com/questions/14243524/cmake-find-library-matching-behavior))

`find_library()` is similar to `find_file()`, but it knows about common library prefixes (e.g. `lib`) and suffixes  (e.g. `.so`).

Reply:
This seems to be working!  In https://ci.appveyor.com/project/guolinke/lightgbm/builds/32185340/job/5tcrah61l6hlv0ab I see

```
Warning message:
In normalizePath(path.expand(path), winslash, mustWork) :
  path[1]="C:/RLibrary/R/lib": The system cannot find the file specified
-- Found LibR: C:/RLibrary/R  
-- LIBR_EXECUTABLE: C:/RLibrary/R/bin/x64/R.exe
-- LIBR_INCLUDE_DIRS: C:/RLibrary/R/include
-- LIBR_CORE_LIBRARY: C:/RLibrary/R/bin/x64/R.dll
```

Don't worry about that "cannot find the file specified" up above those messages, I'm fixing it in #2963

Reply:
Struggling to get this working on Azure. I wrote up this comment bit-by-bit as I tried things, which is how it got so long.

**the problem**:

> We have `"${LIBR_HOME}/bin/${R_ARCH}"` in the set of paths to search, and `LIBR_CORE_LIBRARY` isn't being found...but when I set it directly it works.

When I say "directly", I mean going back to what I originally had

```
set (LIBR_CORE_LIBRARY ${LIBR_HOME}/bin/${R_ARCH}/R.dll CACHE PATH "R core shared library")
```

Let me show you:

* Just adding MinGW to PATH (plus other extra logging that I'll remove):
    - commit: https://github.com/jameslamb/LightGBM/pull/19/commits/05c59f48505c2c1231d3d5e1d2b9059a55602c02
    - AppVeyor: [works](https://ci.appveyor.com/project/jameslamb/lightgbm/builds/32280450/job/vk10elspubecgufx)
    - Azure: [fails](https://dev.azure.com/JayLamb20/jameslamb-lightgbm/_build/results?buildId=389&view=logs&j=b51fa6ee-4364-59e6-4527-dc4aaafd2db2&t=85bde6ad-0953-589b-e1e8-633bb2d9d530)
        - `Could NOT find LibR (missing: LIBR_CORE_LIBRARY)`
* Hardcoding path guess:
    -  commit: https://github.com/jameslamb/LightGBM/pull/19/commits/fdd57243537a125da0e0c176128f6e1beee9a043
    - AppVeyor: [works](https://ci.appveyor.com/project/jameslamb/lightgbm/builds/32280547)
        - you can see log message `LIBR_CORE_LIBRARYwas found by find_library()`
    - Azure: [works](https://dev.azure.com/JayLamb20/jameslamb-lightgbm/_build/results?buildId=391&view=logs&j=b51fa6ee-4364-59e6-4527-dc4aaafd2db2&t=85bde6ad-0953-589b-e1e8-633bb2d9d530)
        - you can see log message `LIBR_CORE_LIBRARY still not found, trying hard-coding`

From the `Get-ChildItem` calls and added cmake logs, I can see in the logs that these are the relevant paths

* AppVeyor
    - R.dll: `C:\projects\lightgbm\RLibrary\R\bin\x64\R.dll`
    - `LIBR_HOME`: `C:/projects/lightgbm/RLibrary/R`
* Azure
    - R.dll: `D:\a\1\s\RLibrary\R\bin\x64\R.dll`
    - `LIBR_HOME`: `D:/a/1/s/RLibrary/R`

So then I thought "ok, the paths are the same...maybe the `CMake` versions are different and `find_library()` behaves differently...they are! and it does!

**CMake versions**

* Appveyor: 3.16
    -  I can see `C:/Program Files (x86)/CMake/share/cmake-3.16` in the logs
* Azure: 3.17
    - following the [Microsoft docs on the 'vs2017-win2016' image we use](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops) to [the list of installed software](https://github.com/actions/virtual-environments/blob/master/images/win/Windows2016-Readme.md#cmake)

We did hear in https://github.com/microsoft/LightGBM/issues/2995#issuecomment-613167751 that our setup wasn't working with `CMake` 3.17.

I found an important-looking line in https://cmake.org/cmake/help/v3.17/release/3.17.html#id17

> When using MinGW tools, the find_library() command no longer finds .dll files by default. Instead, it expects .dll.a import libraries to be available.

I'm confused though, it seems like passing `R.dll`  to `NAMES` should work ([CMake docs](https://cmake.org/cmake/help/latest/command/find_library.html)):

> Each library name given to the NAMES option is first considered as a library file name and then considered with platform-specific prefixes (e.g. lib) and suffixes (e.g. .so). Therefore one may specify library file names such as libfoo.a directly. 

I'm going to try something like this:

```
# look for the core R library
find_library(
  LIBR_CORE_LIBRARY
  NAMES R R.dll
  HINTS "${CMAKE_CURRENT_BINARY_DIR}" "${LIBR_HOME}/lib" "${LIBR_HOME}/bin/${R_ARCH}" "${LIBR_HOME}/bin" "${LIBR_LIBRARIES}"
)

if (WIN32 AND NOT LIBR_CORE_LIBRARY)
    find_file(
        LIBR_CORE_LIBRARY
        NAME R.dll
        HINTS "${CMAKE_CURRENT_BINARY_DIR}" "${LIBR_HOME}/lib" "${LIBR_HOME}/bin/${R_ARCH}" "${LIBR_HOME}/bin" "${LIBR_LIBRARIES}"
    )
endif()
```

I'd rather do that than replace the origin `find_library()` call completely, because I have seen it called different things on some Linux system (like `libR.so` or `R.so`) and I don't want our CMake code to have to know about all the variables CMake uses to store different library prefixes and suffixes.

Reply:
I tried that strategy and it seems like it worked!

* AppVeyor: [works](https://ci.appveyor.com/project/jameslamb/lightgbm/builds/32280807/job/0mw3esec8kaigbcm)
* Azure: [works](https://dev.azure.com/JayLamb20/jameslamb-lightgbm/_build/results?buildId=392&view=logs&j=b51fa6ee-4364-59e6-4527-dc4aaafd2db2&t=85bde6ad-0953-589b-e1e8-633bb2d9d530)
    - _and you can see `Rtools/mingw_64/bin/g++.exe` is being chosen as the compiler_

Reply:
Hmm, so the root cause was in different CMake versions, right? Thanks a lot for your investigation!

> put MinGW compilers on PATH so they're found instead of the ones from chocolatey

I think that more reliable solution will be to use `$env:CXX = path_to_right_compiler`, like we do in the following snippet 
https://github.com/microsoft/LightGBM/blob/0d3e2046533acdda62f15936fb8636f7c8bc27d6/.ci/test.sh#L3-L9
and it is suggested in R installation script
https://github.com/microsoft/LightGBM/blob/0d3e2046533acdda62f15936fb8636f7c8bc27d6/build_r.R#L1-L6

> changed back from "don't run R docs on MinGW builds" to "don't run R docs on Azure"

Sorry, didn't get it. Why did you change it back? I think we can test docs in case of MinGW on Azure.

Reply:
> why did you change it back

Isn't that what you were supporting by https://github.com/microsoft/LightGBM/pull/2936#discussion_r410226345? Maybe I misunderstood.

> I think that more reliable solution will be to use $env:CXX = path_to_right_compiler, like we do in the following snippet

Sure I can do that! But since both chocolatey and MinGW are `g++.exe`, unlike in the example you shared I'll have to do the full path. Like `$env:CXX = "$env:R_LIBS_PATH/Rtools/mingw_64/bin/g++.exe"`.

Let me try that.

Reply:
> Isn't that what you were supporting by #2936 (comment)? Maybe I misunderstood.

That comment was about what I thought **before** CTAN pinning allowed us to move forward. Sorry for confusing you.

The current plan may be the following.
1. In this PR enable full (`--as-cran`) `MINGW` builds at both Appveyor and Azure.
2. In next PR after speeding up VS builds change `MINGW` to `VS` at Azure to cover more use cases.
3. If building time will still take so long, change `--as-cran` to config without docs building in case of `VS`.

Please let me know WDYT about that.

Reply:
> Sure I can do that! But since both chocolatey and MinGW are `g++.exe`, unlike in the example you shared I'll have to do the full path. Like `$env:CXX = "$env:R_LIBS_PATH/Rtools/mingw_64/bin/g++.exe"`.

Yeah, for sure! Just like in 
https://github.com/microsoft/LightGBM/blob/0d3e2046533acdda62f15936fb8636f7c8bc27d6/build_r.R#L4

Reply:
I agree with this plan --> https://github.com/microsoft/LightGBM/pull/2936#discussion_r410927991

Reply:
Ok these changes  worked: https://github.com/jameslamb/LightGBM/pull/19/commits/c145b7dac650e95ebe0491ee6c6d32768fd22dc8

* Azure [works, 13 minutes](https://dev.azure.com/JayLamb20/jameslamb-lightgbm/_build/results?buildId=393&view=logs&j=9f61088a-36ee-57e4-0947-03329fd6ee90&t=2528b315-ee14-54c7-8330-fcfcaf0cbb8e)
* AppVeyor [works, 11 minutes](https://ci.appveyor.com/project/jameslamb/lightgbm/builds/32288119/job/f4uswqd26m6v7uvw)

Added back to this PR in: https://github.com/microsoft/LightGBM/pull/2936/commits/6ca2be8f3c66f16607f655344e51030e4e56c8ec

Reply:
One note...in https://github.com/microsoft/LightGBM/pull/2936/commits/6ca2be8f3c66f16607f655344e51030e4e56c8ec, I removed `Rtools` mingw from `PATH`, so we could be sure that setting `CXX` explicitly was working.  I think that in #2965 we'll have to add it back, because that directory is also where `gendef.exe` is based.

I'll add a link to this comment on this PR.